src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -499,352 +527,305 @@ public Task<Compilation> GetCompilationAsync(SolutionCompilationState compilatio
+                                inProgressState.IntermediateProjects.RemoveAt(0));
+                            this.WriteState(currentState);
+
+                            Contract.ThrowIfTrue(inProgressState.IntermediateProjects.Count > 1 && currentState is not InProgressState);
+                            Contract.ThrowIfTrue(inProgressState.IntermediateProjects.Count == 1 && currentState is not AllSyntaxTreesParsedState);
+                        }
 
-                    while (!intermediateProjects.IsEmpty)
+                        Contract.ThrowIfTrue(currentState is not AllSyntaxTreesParsedState);
+                        return (AllSyntaxTreesParsedState)currentState;
+                    }
+                    catch (Exception e) when (FatalError.ReportAndPropagateUnlessCanceled(e, cancellationToken, ErrorSeverity.Critical))
                     {
-                        cancellationToken.ThrowIfCancellationRequested();
+                        throw ExceptionUtilities.Unreachable();
+                    }
 
-                        // We have a list of transformations to get to our final compilation; take the first transformation and apply it.
-                        var (oldState, action) = intermediateProjects[0];
+                    async Task<(Compilation compilationWithoutGeneratedDocuments, Compilation? staleCompilationWithGeneratedDocuments, CompilationTrackerGeneratorInfo generatorInfo)>
+                        ApplyFirstTransformationAsync(InProgressState inProgressState)
+                    {
+                        Contract.ThrowIfTrue(inProgressState.IntermediateProjects.IsEmpty);
+                        var (oldState, action) = inProgressState.IntermediateProjects[0];
+
+                        var compilationWithoutGeneratedDocuments = inProgressState.CompilationWithoutGeneratedDocuments;
+                        var staleCompilationWithGeneratedDocuments = inProgressState.StaleCompilationWithGeneratedDocuments;
+
+                        // If staleCompilationWithGeneratedDocuments is the same as compilationWithoutGeneratedDocuments,
+                        // then it means a prior run of generators didn't produce any files. In that case, we'll just make
+                        // staleCompilationWithGeneratedDocuments null so we avoid doing any transformations of it multiple
+                        // times. Otherwise the transformations below and in FinalizeCompilationAsync will try to update
+                        // both at once, which is functionally fine but just unnecessary work. This function is always
+                        // allowed to return null for AllSyntaxTreesParsedState.StaleCompilationWithGeneratedDocuments in
+                        // the end, so there's no harm there.
+                        if (staleCompilationWithGeneratedDocuments == compilationWithoutGeneratedDocuments)
+                            staleCompilationWithGeneratedDocuments = null;
```

Comment:
We have a CompilationPair type that we use in some cases here. I'm a bit confused because right now it looks like it's only being used in some partial cases when I'm not sure why we're not using it here. There might be a cleanup available here too (and also FinalizeCompilation...I'm not sure what happened in history here.)

Reply:
Feel free to ignore this bit, this is me thinking out-loud, but if you have a reasonable pause point I might want to try a bit of a refactor but that'd mess you up if you have work in flight.

Reply:
i'm def up for cleaning up compilation pair.  i was also trying to figure out how important it was.